#104 Diary 클래스 재작성

KaquiIOS opened the pull request:
## 개요
기존 Diary class 에서 변경

## 작업사항
1. 기존 Diary Class 이름 변경(->LegacyDiary)
2. String member 에 @NonNull 어노테이션 추가
3. Immutable 하게 작성
4. DB에 저장하는 정보 중 화면에 보이는 아이템이 가져야하는 속성만 추출
5. Emotion Enum Type 추가
resolved #104

KaquiIOS committed:
- feature/issue-104-diary

기존 Diary 클래스 LegacyDiary 클래스로 변경

KaquiIOS committed:
- feature/issue-104-diary

새로운 Diary class 작성
1. String 타입 멤버에 @NonNull 어노테이션 추가
2. Immutable 하게 작성

KaquiIOS committed:
- feature/104-issue-diary

리팩토링 기능 사용중 SQL 에 있는 값들까지 잘못 변경된 점 수정

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,36 @@
-    // 설정한 태그들
-    @ColumnInfo(name = "tags")
+    @NonNull
     private final String tags;
```
List<String>이어야 하지 않을까요?

zeropol2 replied:
필수가 아니라면 @Nullable 추가하시죠

KaquiIOS replied:
List 타입으로 가지고 있지않고 String 형태로 저장해뒀다가 #을 기준으로 split 하여 가져오는 방식으로 작성하기로 했었습니다.  List 로 가지고 있는게 더 나을까요 ?

zeropol2 replied:
넵 List가 더 나을 것 같아요

KaquiIOS replied:
수정했습니다 :-)
https://github.com/boostcampth/boostcamp3_H/pull/106/commits/afe3394d8447b6b4371847b32d0019fd926a30c0

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,36 @@
-    // 선택한 감정 번호
-    @ColumnInfo(name = "selectedEmotion")
+    private final int id;
     private final int selectedEmotion;
```
사용자가 선택한 emotion과 api에 의해서 분석된 emotion 두가지를 다 들고 있어야 하는건 아닌가요?
화면에 어떻게 그려지는지 몰라서요~ 화면에는 사용자가 선택한 emotion만 보여지나요?

KaquiIOS replied:
화면에 그려지는 이모티콘은 사용자 선택 이모티콘만 보여지고 통계를 보여줄 때 분석된 감정을 선택한 감정과 같이 보여주게 됩니다

zeropol2 replied:
아하 이건 개발과는 상관없지만, 두개 다 보여주는게 더 재밌지 않을까요?

KaquiIOS replied:
이 부분은 팀원들과 이야기를 해보고 추가하면 알려드리도록 하겠습니다 :-)

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -90,49 +53,4 @@ public String getRecordFilePath() {
     public int getSelectedEmotion() {
         return selectedEmotion;
     }
```
recordDate, recordFilePath, tags도 getter가 있어야 하지 않을까요?

KaquiIOS replied:
접혀있는 부분에 getter 들이 존재합니다 !

MeongByeol replied:
접힌 부분에 getter 가 존재합니다 !

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,36 @@
+ * Diary class 새로 작성
+ * 기존 Diary 는 DiaryEntity 의 모습과 동일하고,
+ * 화면에서 보이는 일기들이 가져야하는 속성들만 빼서 작성
+ * 최대한 Immutable 하게 작성 */
```
불필요한 주석일 것 같네요

KaquiIOS replied:
주석은 지우도록 하겠습니다 !

KaquiIOS replied:
주석 삭제했습니다
https://github.com/boostcampth/boostcamp3_H/pull/106/commits/afe3394d8447b6b4371847b32d0019fd926a30c0

KaquiIOS committed:
- feature/issue-104-diary

피드백 주신 부분들을 수정하였습니다.
1. tag type 변경 (String -> List<String>, Nullable 어노테이션)
2. 불필요한 주석 제거

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,33 @@
-                 final int selectedEmotion,
-                 final int analyzedEmotion,
-                 final long timeStamp) {
+                 @NonNull List<String> tags,
```
이거 필드 선언에는 Nullable인데 어떤게 맞는건가요? tag입력을 안하고 일기를 저장할수 있다? 없다?

KaquiIOS replied:
네 태그는 없어도 괜찮다고 팀원들과 협의했습니다 !

KaquiIOS replied:
getter 및 매개변수 어노테이션을 변경했습니다
https://github.com/boostcampth/boostcamp3_H/pull/106/commits/840a1551cdd4015e2889967d11a6a873cd287106

KaquiIOS committed:
- feature/issue-104-diary

getter 의 annotation 수정을 빼먹었습니다 ㅠㅠ 다음부턴 더 꼼꼼하게
보도록 하겠습니다.

KaquiIOS committed:
- feature/issue-104-diary

- getter 및 매개변수 어노테이션을 변경했습니다.

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,33 @@
+    @Nullable
+    private final List<String> tags;
+    private final int id;
     private final int selectedEmotion;
```
emotion을 int로 저장하시는 대요~ 좀 설명해 주실수 있나요?

KaquiIOS replied:
API 에서는 7가지 감정을 List(String) 타입으로 반환합니다 !
그 감정들은 fear, mad, sad, frustrated, neutral, good, happy, excited 로 구성되는데 분석된 감정들을 좋은(1)/평범(0)/나쁨(-1) 카테고리로 묶어 점수를 부여하게 됩니다. 그래서 그 점수들을 계산해서 0-4 사이의 값으로 변경하여 사용합니다. 
0 - 제일 안좋음
1 - 안좋음
2 - 평범
3 - 좋음
4 - 완전 좋음
String 으로 처리를 하려고 했지만 통계에서 감정을 구하는 부분에서 index 로 접근하는게 더 간편하다는 판단하에 int 타입으로 설정했습니다.

< 변환 과정 >
     * 분석된 감정은 0 - 4 사이 값으로 매핑된다
     *  알고리즘은 API 에서 분석한 감정들을 안좋은 감정, 보통, 좋은 감정 3가지로 나눈후
     *  안좋은 감정 -1 점, 보통 감정 0 점, 좋은 감정 1 점으로 설정하고
     *  (각 감정 X 점수들을 합산) / 전체 시간 = -1 ~ 1 사이의 값으로 정규화하고
     *  0.4 단위로 각 구간을 0 - 4 값으로 매핑하여 돌려주는 것이다.

zeropol2 replied:
그렇다면 요 값들을 enum으로 만들어서 관리하는게 좋겠네요.
```
enum Emotion {
        GOOD(0), VERY_GOOD(1), BAD(2);
        
        private int value;
        
        Emotion(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }
    }
```
요런 식으로요~

KaquiIOS replied:
Enum 이 좋을까요 IntDef 가 좋을까요 ? 일단은 Enum 으로 하겠습니다 !

zeropol2 replied:
넵 enum 가시죠~

KaquiIOS replied:
enum type 에도 @NonNull 어노테이션을 달 수 있는지 찾아보다 좀 늦었습니다 ! 수정했습니다 ~
https://github.com/boostcampth/boostcamp3_H/pull/106/commits/58461840b07d9e3e69e3b255dfa91bc651ef587d

KaquiIOS committed:
- feature/issue-104-diary

- int 로 표현된던 emotion 값을 Enum 으로 설정

zeropol2 reviewed (approved):
LGTM 머지하고 다음작업 가시죠~

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,33 @@
+    @Nullable
+    private final List<String> tags;
+    private final int id;
+    private final Emotion selectedEmotion;
```
앗 잠시만요 여기도 어노테이션 챙겨주세요

KaquiIOS replied:
수정했습니다 !
https://github.com/boostcampth/boostcamp3_H/pull/106/commits/182650b9ef20ebbfc46c7dffc3ec18d9184acbf6

KaquiIOS replied:
추가했습니다 ~
https://github.com/boostcampth/boostcamp3_H/pull/106/commits/182650b9ef20ebbfc46c7dffc3ec18d9184acbf6

KaquiIOS committed:
- feature/issue-104-diary

어노테이션 추가

zeropol2 commented on app/src/main/java/teamh/boostcamp/myapplication/data/model/Diary.java:
```diff
@@ -1,73 +1,34 @@
-    // 저장한 날
-    @ColumnInfo(name = "recordDate")
     @NonNull
     private final String recordDate;
```
제가 이부분 놓쳤는데, 이건 String이 아니라 Date내지는 long으로 저장해야 할것 같은데요??
